clipper_admin/clipper_admin/docker/docker_container_manager.py

```diff
@@ -115,22 +125,43 @@ def start_clipper(self,
             msg = "Unable to Connect to Docker. Please Check if Docker is running."
             raise ClipperException(msg)
 
+        containers_in_cluster = self.docker_client.containers.list(
+            filters={
+                'label':
+                ['ai.clipper.container.label={}'.format(self.cluster_name)]
+            })
+        if len(containers_in_cluster) > 0:
+            raise ClipperException(
+                "Cluster {} cannot be started because it already exists. "
+                "Please use clipper_conn.connect() to connect to it.".format(
+                    self.cluster_name))
+
         if not self.external_redis:
             logger.info("Starting managed Redis instance in Docker")
+            self.redis_port = find_unbound_port(self.redis_port)
+            redis_labels = self.common_labels.copy()
+            redis_labels[CLIPPER_DOCKER_PORT_LABELS['redis']] = str(
+                self.redis_port)
```

Comment:
TL;DR: useful for ClipperConnection.connect()

Because the port is now floating, we no longer have knowledge of port mapping. So when user calls connect, we will know which port correspond to which container.

There will an alternative by just using docker python client but requires a "three way join" among docker containers, docker network, and some docker client magics; current approach is more straightforward to implement and understand.